crates/sui-types/src/crypto.rs

```diff
@@ -1,205 +1,109 @@
 // Copyright (c) 2022, Mysten Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
-use crate::base_types::{decode_bytes_hex, AuthorityName, SuiAddress};
+use crate::base_types::{AuthorityName, SuiAddress, ToAddress};
 use crate::committee::{Committee, EpochId};
 use crate::error::{SuiError, SuiResult};
 use crate::sui_serde::Base64;
 use crate::sui_serde::Readable;
 use crate::sui_serde::SuiBitmap;
-use anyhow::anyhow;
 use anyhow::Error;
 use base64ct::Encoding;
 use digest::Digest;
-use ed25519_dalek as dalek;
-use ed25519_dalek::{Keypair as DalekKeypair, Verifier};
-use narwhal_crypto::ed25519::{Ed25519KeyPair, Ed25519PublicKey};
-use once_cell::sync::OnceCell;
+use narwhal_crypto::ed25519::{
+    Ed25519KeyPair, Ed25519PrivateKey, Ed25519PublicKey, Ed25519PublicKeyBytes,
+};
+pub use narwhal_crypto::traits::KeyPair as KeypairTraits;
+pub use narwhal_crypto::traits::{
+    AggregateAuthenticator, Authenticator, SigningKey, ToFromBytes, VerifyingKey,
+};
+use narwhal_crypto::Verifier;
 use rand::rngs::OsRng;
 use roaring::RoaringBitmap;
 use schemars::JsonSchema;
 use serde::{Deserialize, Serialize};
 use serde_with::serde_as;
 use serde_with::Bytes;
 use sha3::Sha3_256;
-use std::borrow::Borrow;
 use std::collections::HashMap;
 use std::hash::{Hash, Hasher};
-use std::str::FromStr;
 
-// TODO: Make sure secrets are not copyable and movable to control where they are in memory
-#[derive(Debug)]
-pub struct KeyPair {
-    key_pair: DalekKeypair,
-    public_key_cell: OnceCell<PublicKeyBytes>,
-}
+// Comment the one you want to use
+pub type KeyPair = Ed25519KeyPair; // Associated Types don't work here yet for some reason.
```

Comment:
Do we want to merge this first then allow crypto agility? I.e., eventually the keypair can be of any supported type (ECDSA, BLS, EdDSA etc)

Reply:
It also doesn't harm on the future to have two key types (one for users and another for validators), mostly because we'll allow different schemes per type (ie we don't expect validators to ever support ECDSA)

Reply:
I have another PR that will address this!